feat: Architecture page (#777)
appsmithorg/appsmith-docs: website/docs/reference/appsmith-architecture/README.md

> # Appsmith Architecture

Reviewer: ```suggestion
# Architecture
```

# Appsmith Architecture

> Getting started on the Appsmith platform is quick and secure, available via the [cloud](http://app.appsmith.com/user/sign-up) as well as [self-hosted](/getting-started/setup#self-hosted) instances. Both versions are extremely similar in features and architecture, deployed as a single Docker container with a single volume for storing persistent data. With a [single line of code](/getting-started/setup/installation-guides/docker#setup-with-docker-run) or a one-click interaction via the AWS or DigitalOcean Marketplace or Heroku, Appsmith can be installed and put to use.

Author: Ah good catch, I misunderstood the list in the source doc, makes better sense now. Fixed

Updated the intro; didn't change the content of it much, but removed a bit and pivoted the tone so it just tells the user 1) how Appsmith is hosted, 2) the two versions (CE/BE), 3) and primes them for reading the page.

I think it reads better and more on-topic, did it turn out okay? Let me know, I'm not sure how else to word the intro :)

> Getting started on the Appsmith platform is quick and secure, available via the [cloud](http://app.appsmith.com/user/sign-up) as well as [self-hosted](/getting-started/setup#self-hosted) instances. Both versions are extremely similar in features and architecture, deployed as a single Docker container with a single volume for storing persistent data. With a [single line of code](/getting-started/setup/installation-guides/docker#setup-with-docker-run) or a one-click interaction via the AWS or DigitalOcean Marketplace or Heroku, Appsmith can be installed and put to use.

Reviewer 2: looks good :)

Read further to learn about how the platform is organized, built, and deployed.

## Self-hosted instance architecture

> The diagram below illustrates how the different pieces of the platform relate to each other:

Reviewer 2: ```suggestion
The diagram below illustrates the different components and how these services communicate with each other.
```

![Appsmith Deployment Architecture](/img/Appsmith_Deployment_Architecture.png)

## How it works

> A single Docker container runs the following processes to deploy and get started with Appsmith:

Author: This should apply in all cases; Appsmith is built within a Docker container which runs all of the processes in the list. There are various methods of deploying / hosting the container(s), however it's always in the container.

> A single Docker container runs the following processes to deploy and get started with Appsmith:

Reviewer: That's right.

> 1. The **backend server** as a Java process, which handles:

Reviewer 2: Use the names as shown in the diagram so that users are not confused

>     1. Authentication: username and password, OAuth 2.0 with Google and GitHub, SSO with OIDC and SAML

Author: - Updated to `login credentials`
- added links on Google, Github, OIDC, and SAML to their relevant `Instance Config > Auth > ...` pages

    2. A CRUD API for users, workspaces, applications, pages, widgets

>     3. An action execution module that executes your queries to your databases and APIs

Author: I agree there, I think it describes the structure of the platform well without getting too deep into implementation

4. The **MongoDB server**, which stores all persistent data needed and generated by Appsmith’s building blocks, like:
    1. Your users’ details and permissions
    2. Workspaces you have created
    3. Applications you are building and have deployed
    4. Datasources you have connected to
    5. Queries that connect those datasources to your apps.

5. The **Redis server**, which is used for storing user sessions and cache permissions that a user has access to directly and indirectly.

## Process management

To manage the processes needed to function, Appsmith uses [Supervisor](http://supervisord.org/). This is a lightweight process manager that, when given a list of things to run, runs them, restarts them if they crash, and collects their logs for maintenance and debugging.

When you run the Docker command to start the Appsmith container, it identifies which processes are required to start and run. This includes the backend server, the RTS, and the NGINX server -- as well as any additional processes that haven't been plugged in by the user externally, such as a custom MongoDB or Redis instance.

> For each of these processes, the container’s entry-point script (the one that Docker invokes when the container starts) generates a config file with all the details needed for supervisord. For example, the config file generated for the backend Java server includes the full set of commands for starting the process, as well as all necessary environment variables and settings. This same entry-point script also helps set up SSL certificates and install custom CA roots.

Reviewer: Can we add a cross-reference to the `supervisord`.

Author: Added link to appsmith Supervisor page

Reviewer: While reading it again, I feel the flow is a bit broken. We are introducing the utility at line 47 but using it before at line 45 too. Let's bring the intro for Supervisord before using it at line 47.

Author: Looks like my understanding of the different pieces of Supervisor wasn't great, I looked through its docs some more and updated this section accordingly, I think it might flow a bit better now

Supervisord also comes with a command line tool that you can use to control these processes individually. For example, you can run `supervisorctl restart redis` if you want to restart the Redis process. Or, as another example, the `Save and Restart` button under `Admin Settings` on the Appsmith UI is powered the `supervisorctl` tool; clicking that button after you make changes with the UI prompts the backend server to execute `supervisorctl restart backend`, which restarts the backend process.

> ### Docker

Author: Removed most of the mentioned Docker and Kubernetes subheadings -- you're right, they were instructing the user on deploying their own instance rather than talking about how Appsmith itself is structured. The information in these sections already exists on other pages, so it was removed.

I combined a bit from each of those into a new heading **Deployment** that informs the user how the Appsmith Cloud is deployed.

> ### Docker

Reviewer 2: Do we need to talk about deployment under Architecture?

> ### Docker

Reviewer: IMO, the content under Deployment head can be a part of How it works itself. Thoughts??

> ### Docker

Reviewer: @mohanarpit - Could you please suggest an alternative here?

Appsmith runs as a single Docker container, housing all five processes in one image via Supervisord.

As described above in [process management](#process-management), this configuration provides flexibility in how you set up your Appsmith instance. Appsmith can take care of the entire batch of processes on its own, or you can choose to [drop in your own custom elements](/getting-started/setup/instance-configuration/custom-mongodb-redis) like MongoDB or Redis by modifying environment variables.

### Kubernetes

Kubernetes is an open source software used to manage and scale wide deployments of applications in containers; Appsmith Cloud runs on geographically distanced Kubernetes clusters via AWS. This deployment strategy allows the Appsmith Cloud platform to be scale and avoid downtime during outages, as well as conduct rolling updates with minimal disruptions to service.

> If you choose to [deploy Appsmith on Kubernetes](https://docs.appsmith.com/getting-started/setup/installation-guides/kubernetes), the only solution is to use our [official Helm chart](https://helm.appsmith.com). This Helm chart manages MongoDB and Redis as separate [Kubernetes pods](https://kubernetes.io/docs/concepts/workloads/pods/) so your Appsmith instances can scale horizontally. As before, this method also allows you the option to use external MongoDB and Redis instances that you manage on your own.

Reviewer: Can we have relative paths linked instead of the URL for internal links?

Author: Oops!
